synapse/handlers/directory.py

```diff
@@ -323,18 +303,36 @@ def _update_canonical_alias(self, requester, user_id, room_id, room_alias):
             room_id, EventTypes.CanonicalAlias, ""
         )
 
+        # There is no canonical alias, nothing to do.
+        if not alias_event:
+            return
+
+        # An updated canonical alias event must be sent if the removed alias was
+        # set as the alias on the canonical_alias event or listed in the
+        # alt_aliases field.
         alias_str = room_alias.to_string()
-        if not alias_event or alias_event.content.get("alias", "") != alias_str:
+        alt_aliases = alias_event.content.get("alt_aliases", [])
```

Comment:
we generally need to be a bit careful when reading values out of event bodies: in this case, there's nothing saying that `content["alt_aliases"]` will definitely be a list of strings. (It should be, if all users everywhere are behaving themselves... but users.)

Synapse's code base is littered with places that we don't do this properly, but: we either need to check that `alt_aliases` is actually a list before trying to treat it as one, or otherwise convince ourselves that it's not going to explode in a ball of flames.

Reply:
I was wondering about, but figured I'd keep it simple at first. I'll add some more error checking code here.